floors/floors.go

```diff
@@ -0,0 +1,203 @@
+	enforceRateMax   int    = 100
+)
+
+// EnrichWithPriceFloors checks for floors enabled in account and request and selects floors data from dynamic fetched floors JSON if present
+// else selects floors JSON from req.ext.prebid.floors and update request with selected floors details
+func EnrichWithPriceFloors(bidRequestWrapper *openrtb_ext.RequestWrapper, account config.Account, conversions currency.Conversions) []error {
+
+	if bidRequestWrapper == nil || bidRequestWrapper.BidRequest == nil {
+		return []error{errors.New("Empty bidrequest")}
+	}
+
+	if isPriceFloorsDisabled(account, bidRequestWrapper) {
+		return []error{errors.New("Floors feature is disabled at account level or request")}
+	}
+
+	floors, err := resolveFloors(account, bidRequestWrapper, conversions)
+
+	updateReqErrs := updateBidRequestWithFloors(floors, bidRequestWrapper, conversions)
+	updateFloorsInRequest(bidRequestWrapper, floors)
+	return append(err, updateReqErrs...)
+}
+
+// updateBidRequestWithFloors will update imp.bidfloor and imp.bidfloorcur based on rules matching
+func updateBidRequestWithFloors(extFloorRules *openrtb_ext.PriceFloorRules, request *openrtb_ext.RequestWrapper, conversions currency.Conversions) []error {
+	var (
+		floorErrList []error
+		floorVal     float64
+	)
+
+	if extFloorRules == nil || extFloorRules.Data == nil || len(extFloorRules.Data.ModelGroups) == 0 {
+		return []error{}
+	}
+
+	modelGroup := extFloorRules.Data.ModelGroups[0]
+	if modelGroup.Schema.Delimiter == "" {
+		modelGroup.Schema.Delimiter = defaultDelimiter
+	}
+
+	extFloorRules.Skipped = new(bool)
+	if shouldSkipFloors(modelGroup.SkipRate, extFloorRules.Data.SkipRate, extFloorRules.SkipRate, rand.Intn) {
+		*extFloorRules.Skipped = true
+		return []error{}
+	}
+
+	floorErrList = validateFloorRulesAndLowerValidRuleKey(modelGroup.Schema, modelGroup.Schema.Delimiter, modelGroup.Values)
+	if len(modelGroup.Values) > 0 {
+		for i, imp := range request.GetImp() {
+			desiredRuleKey := createRuleKey(modelGroup.Schema, request.BidRequest, request.Imp[i])
+			matchedRule, isRuleMatched := findRule(modelGroup.Values, modelGroup.Schema.Delimiter, desiredRuleKey, len(modelGroup.Schema.Fields))
+
+			floorVal = modelGroup.Default
+			if isRuleMatched {
+				floorVal = modelGroup.Values[matchedRule]
+			}
+
+			floorMinVal, floorCur, err := getMinFloorValue(extFloorRules, request.Imp[i], conversions)
+			if err == nil {
+				floorVal = math.Round(floorVal*10000) / 10000
+				bidFloor := floorVal
+				if floorMinVal > 0.0 && floorVal < floorMinVal {
+					bidFloor = floorMinVal
+				}
+
+				if bidFloor > 0.0 {
+					imp.BidFloor = math.Round(bidFloor*10000) / 10000
+					imp.BidFloorCur = floorCur
+				}
+				if isRuleMatched {
+					updateImpExtWithFloorDetails(imp, matchedRule, floorVal, imp.BidFloor)
+				}
+			} else {
+				floorErrList = append(floorErrList, fmt.Errorf("Error in getting FloorMin value : '%v'", err.Error()))
+			}
+		}
+		err := request.RebuildImp()
```

Comment:
Why are we doing a rebuild at this point? Is there some code further along that is accessing the raw request rather than the wrapper and thus needs to see the modified values in the raw request? If so, then I would prefer fixing that code to use the wrapper rather than refactoring the wrapper to work around that deficiency in the code. Or at worst, moving the rebuild to the start of the code that depends on the raw request being up to date so it is only rebuilt if and when necessary.

Reply:
This is done to update impression extension with floors details, which shall be used during floors enforcement to update bid extension about floor rule details.
Rather than rebuilding complete request, only request fields are updated and respective component is rebuild.